### ISSUE TITLE
Make it easy to perform retries on requests - error results should include relevant info like request url, headers, etc.

### ISSUE THREAD

**Issue by sreeise** [sreeise/graph-rs#363]
The client does not have a way to handle 429 responses (too many requests) that results from throttling. There should be a way to retry the requests that failed.

- It is difficult handle retries.
- `GraphRequest<T`>` cannot be recreated in anyway.
- `GraphResponse<T>` has information necessary for a retry in the event of a throttle occurring. However because this is a 429 error response the api client returns a `GraphResult` with an error object and this does not provide any relevant information to make a retry request.
- `GraphResponse` does not have relevant information for retrying a request.

**Comment by JoschuaL** [comment]
So one thing i can be up with that might be a reasonable way to handle this would be a send_with_retry function implemented on IntoResponse (analogue to the send function) that could be configured how and when to retry a request. The configuration could be done through a builder pattern struct that would contain the number of retires, a list of status codes to retry on, and potentially a Fn(GraphResult) -> bool (or maybe GraphResponse or even T) with a default implementation that just retries 1 or 2 times on any GraphError. I'll see if I can code a pull request together, though I'm not promising anything.

(Sorry for the double post on the discussion)

**Comment by sreeise** [comment]
> So one thing i can be up with that might be a reasonable way to handle this would be a send_with_retry function implemented on IntoResponse (analogue to the send function) that could be configured how and when to retry a request. The configuration could be done through a builder pattern struct that would contain the number of retires, a list of status codes to retry on, and potentially a Fn(GraphResult) -> bool (or maybe GraphResponse or even T) with a default implementation that just retries 1 or 2 times on any GraphError. I'll see if I can code a pull request together, though I'm not promising anything.
> 

For sure. If you want to get something up just so that we can look at what you had in mind, that works as well.

**Comment by JoschuaL** [comment]
Ok, so I got some very early PoC that is flexible to a degree that might not be needed, and I have no idea how ergonomic and performant it would be. But its a start.

```rust

pub struct RetryConfigBuilder<T> {
    // Number of retries
    times: usize,
    // Timeout between retries, either derived from a fixed value or from GraphResult
    fixed_timeout: time::Duration,
    timeouts: Vec<Box<dyn Fn(&GraphResult<GraphResponse<T>>) -> Option<time::Duration>>>,
    // Status Codes on which to retry unconditionally
    codes: BTreeSet<StatusCode>,
    // Conditional Retries for results and errors
    result_validators: Vec<Box<dyn Fn(&GraphResponse<T>) -> bool>>,
    error_validators: Vec<Box<dyn Fn(&GraphFailure) -> bool>>,
}

//Manual Builder Pattern for tighter control
impl<T> RetryConfigBuilder<T> {
    pub fn num_retries(mut self, retries: usize) -> Self {
        self.times = retries;
        self
    }

    pub fn fixed_timeout(mut self, timeout: time::Duration) -> Self {
        self.fixed_timeout = timeout;
        self
    }

    pub fn timeout(
        mut self,
        timeout: Box<dyn Fn(&GraphResult<GraphResponse<T>>) -> Option<time::Duration>>,
    ) -> Self {
        self.timeouts.push(timeout);
        self
    }

    pub fn response_codes(mut self, mut codes: BTreeSet<StatusCode>) -> Self {
        self.codes.append(&mut codes);
        self
    }

    pub fn result_validator(mut self, validator: Box<dyn Fn(&GraphResponse<T>) -> bool>) -> Self {
        self.result_validators.push(validator);
        self
    }

    pub fn error_validator(mut self, validator: Box<dyn Fn(&GraphFailure) -> bool>) -> Self {
        self.error_validators.push(Box::new(validator));
        self
    }

    pub fn build(self) -> RetryConfig<T> {
        RetryConfig {
            times: self.times,
            fixed_timeout: self.fixed_timeout,
            timeouts: self.timeouts,
            codes: self.codes,
            result_validators: self.result_validators,
            error_validators: self.error_validators,
        }
    }

    pub fn new() -> Self {
        Self {
            times: 3,
            fixed_timeout: Duration::ZERO,
            timeouts: Vec::new(),
            codes: BTreeSet::new(),
            result_validators: Vec::new(),
            error_validators: Vec::new(),
        }
    }

    // Example. Should suffice to handle Throttling on Graph Requests
    pub fn new_throttle_retry() -> Self {
        let mut codes = BTreeSet::new();
        codes.insert(StatusCode::TOO_MANY_REQUESTS);
        Self {
            times: 3,
            fixed_timeout: Duration::ZERO,
            timeouts: vec![Box::new(|gr: &GraphResult<GraphResponse<T>>| {
                if let Err(GraphFailure::GraphError(e)) = gr {
                    if e.code == StatusCode::TOO_MANY_REQUESTS {
                        if let Some(headers) = e.headers.clone() {
                            if let Ok(value) = serde_json::to_value(headers) {
                                return value["Retry-After"]
                                    .as_u64()
                                    .map(time::Duration::from_secs);
                            }
                        }
                    }
                }
                None
            })],
            codes,
            result_validators: Vec::new(),
            error_validators: Vec::new(),
        }
    }
}

impl<T> Default for RetryConfigBuilder<T> {
    fn default() -> Self {
        Self::new()
    }
}
pub struct RetryConfig<T> {
    times: usize,
    fixed_timeout: time::Duration,
    timeouts: Vec<Box<dyn Fn(&GraphResult<GraphResponse<T>>) -> Option<time::Duration>>>,
    codes: BTreeSet<StatusCode>,
    result_validators: Vec<Box<dyn Fn(&GraphResponse<T>) -> bool>>,
    error_validators: Vec<Box<dyn Fn(&GraphFailure) -> bool>>,
}

impl<T> RetryConfig<T> {
    // derives timeout from First function returning a value, or fixed timeout otherwise
    pub fn get_timeout(&self, res: &GraphResult<GraphResponse<T>>) -> time::Duration {
        for timeout in self.timeouts.iter() {
            if let Some(time) = timeout(res) {
                return time;
            }
        }
        self.fixed_timeout
    }

    fn validate(&self, res: &GraphResult<GraphResponse<T>>) -> bool {
        match res {
            Ok(resp) => {
                for validator in self.result_validators.iter() {
                    if !validator(resp) {
                        return false;
                    }
                }
            }
            Err(e) => {
                for validator in self.error_validators.iter() {
                    if !validator(e) {
                        return false;
                    }
                }
            }
        }
        true
    }

    // retries if validation fails or on hitting one of the configured codes.
    pub fn get_retry(&self, res: &GraphResult<GraphResponse<T>>) -> bool {
        !self.validate(res)
            || match res {
                Err(GraphFailure::GraphError(e)) => self.codes.contains(&e.code),
                Ok(res) => self.codes.contains(&res.status()),
                _ => false,
            }
    }
}

impl<'a, T> IntoResponseBlocking<'a, T>
where
    for<'de> T: serde::Deserialize<'de>,
{
    pub fn build(self) -> DispatchBlocking<T> {
        let builder = self.client.build();
        DispatchBlocking::new(builder, None, self.error)
    }

    pub fn send(self) -> GraphResult<GraphResponse<T>> {
        if self.error.is_some() {
            return Err(self.error.unwrap_or_default());
        }
        self.client.execute()
    }

    // Example how the implementation might look. Configuration could potentially also be moved to the client.
    pub fn send_retry(self, config: RetryConfig<T>) -> GraphResult<GraphResponse<T>> {
        if self.error.is_some() {
            return Err(self.error.unwrap_or_default());
        }
        let mut res: GraphResult<GraphResponse<T>> = self.client.execute();
        for _ in 0..config.times {
            if config.get_retry(&res) {
                thread::sleep(config.get_timeout(&res));
                res = self.client.execute();
            } else {
                break;
            }
        }
        res
    }
}
```

**Comment by sreeise** [comment]
Also I think you already figured it out but just in case the headers are in the graph-error project and it just uses a HeaderMap from the reqwest crate which is pretty easy to use. 
https://github.com/sreeise/graph-rs/blob/3740b4db4bbd0af23f74fadf74b2dd79c13de6c6/graph-error/src/headers.rs#L35-L42

TBH the headers part could use a change from the error crate to the http crate but that may be a change for a later time.

**Comment by sreeise** [comment]
> ```rust
>             if config.get_retry(&res) {
>                thread::sleep(config.get_timeout(&res));
>                res = self.client.execute();
>            }
> ```

We probably shouldn't block the main thread here, unless that is wanted, but I am guessing you did this just as an example.

>```rust
>    timeouts: Vec<Box<dyn Fn(&GraphResult<GraphResponse<T>>) -> Option<time::Duration>>>,
>    codes: BTreeSet<StatusCode>,
>    result_validators: Vec<Box<dyn Fn(&GraphResponse<T>) -> bool>>,
>    error_validators: Vec<Box<dyn Fn(&GraphFailure) -> bool>>,
>```

Are these wrapped in a box for a specific reason?  Meaning other then to store the `Fn`.

>```rust
>            if config.get_retry(&res) {
>```


I feel like this could be done without passing the full result to the retry config.  If the the methods accepted specific values like the status code it would actually help prevent errors. For instance, it may not be an error that is returned from Microsoft because there was a failure somewhere in the chain before the request was ever sent. The error could be multiple different things such as badly formed JSON in a POST request. 

There are also a few things that we have to keep in mind if the client were to keep track of the retry after time. First, the access token could be expired. For this to work we would need to assume that the `Retry-After` header must be in the headers for a retry to occur and that if the header is there then that means the access token is still good. Typically with 429 status codes, `Retry-After` headers are included. However, if this is not the case we potentially could end up in a circular pattern where the client keeps making the retry request with an expired access token. This is less likely of a scenario though. I didnt see any issues with how this was written that would make me think that we would be more likely to get stuck in a loop here but we should still make sure on that. 

And from what I can tell, you solved this by checking the status code and only doing a retry on those results that have a 429 status code. But there is also the issue where an access token could be expired by the time we call a retry request (after waiting the specified time in the `Retry-After` header).**

>```rust
>                 for validator in self.error_validators.iter() {
>                    if !validator(e) {
>                        return false;
>                    }
>                }
>```

> ```rust
>    // retries if validation fails or on hitting one of the configured codes.
>    pub fn get_retry(&self, res: &GraphResult<GraphResponse<T>>) -> bool {
>        !self.validate(res)
>            || match res {
>                Err(GraphFailure::GraphError(e)) => self.codes.contains(&e.code),
>                Ok(res) => self.codes.contains(&res.status()),
>                _ => false,
>            }
>```


Doesn't seem to be clear what the `validator` method is doing? What I mean is that the validation of status codes and the Retry-After header are happening elsewhere and not in the validator so it seems odd. And I am not sure there actually needs to be an outside validator unless your wanting to perform retries based on something unknown to the client itself (retries not based on getting throttled)?

Overall I think I would like to see more of what this would look like and how it would work within the client. My initial thought was that we could provide a way to rebuild the request so that the caller can make the api request whenever the throttle time is expired and also get access to the headers and other response values when a Result<Error> is returned. 

You have provided an alternate solution which I think could work but I am just not sure how it fits into the current ecosystem. But I like the idea of a retry config. I have been wanting to make the client and/or the way requests are made more configurable so this is something that is going to help with that if it works out. Possibly a small working example might be helpful if you wanted to do that. But that is completely up to you and I appreciate the feedback and contribution that you have made already just by pointing out the issues with retrying a request after getting throttled.

**Comment by JoschuaL** [comment]
Yeah, blocking the thread was just meant as an example.

I wanted the validators to allow configuring a retry on arbitrary conditions. I'm not sure it would be needed in practice as configuring the status codes could already be enough, but i wanted to keep options open.

Since im passing the GraphResult. If a GraphResult can be coming from somewhere other than graph that would be kind of unintuitive. If thats the case then i guess some other way to pass the responses from graph to the validators would be needed.

And yes, this is meant for very configurable retries outside of throtteling, with throtteling just being one of the possible defaults. 

Of course checking if the token is still valid before a retry would be possible, though since we are typically dealing throttling for only few seconds I'm not sure it would just be easier to retry and just get the expired token response from graph (as that has a different error code) and i'm not sure shortcutting that 1 request would really be worth it.

**Comment by sreeise** [comment]
> Yeah, blocking the thread was just meant as an example.
> 
> I wanted the validators to allow configuring a retry on arbitrary conditions. I'm not sure it would be needed in practice as configuring the status codes could already be enough, but i wanted to keep options open.
> 
> Since im passing the GraphResult. If a GraphResult can be coming from somewhere other than graph that would be kind of unintuitive. If thats the case then i guess some other way to pass the responses from graph to the validators would be needed.
>

Yeah I think im starting to lean toward needing a better end result value. Like having an error field on GraphResponse and just returning that with all the request info.  

I think the retry config might be better as a defined trait that the sdk user/caller could implement. It would provide better control over what happens in some cases but still provide things like a manual builder implementation. Still want to allow the validators to retry based on any condition it wants, but if the user just wants to match on status codes then there is no reason to require that they implement a trait or anything which is where the builder would come in. So im really just thinking that having both a way to let the caller use retry in a simple and common way and with the ability to get more advanced would be ideal.

**Comment by sreeise** [comment]
I apologize I never got back to you on this. I am going to close this one out since major changes were released in version 1.0.0 that make it so the changes we talked about less applicable. If you get a chance to use version 1.0.0 and still have a feature request please do reopen this issue or create a new issue with the feature request you would like.
